docs/blog/posts/constitutional-ai-or-validation.md

```diff
@@ -0,0 +1,221 @@
+
+Tools like Constitutional AI address shortcomings by using AI feedback to evaluate outputs. In this approach, an AI system provides a set of principles for making judgments about generated text. These principles guide the model's behavior. Essentially, it's a form of validation. Instead of manually checking against a set of rules, we now have a model that can perform the validation. For example, a content moderation system could use a model to automatically ban certain keywords. Previously, an error might have been thrown upon detection, but now we have a self-correcting system.
+
+This post will explore how we can achieve this using Pydantic and Instructor without introducing new standards or terms for validation.
+
+## Validation in Software 1.0
+
+Common types of validation in Software 1.0 include:
+
+-  Type checking
+-  Range checking
+-  Length checking
+
+Here's an example model that uses Pydantic to validate the range and length of a string, taken directly from the Pydantic documentation:
+
+```python
+from pydantic import BaseModel, ValidationError, field_validator
+
+class UserModel(BaseModel):
+    id: int
+    name: str
+
+    @field_validator('name')
+    @classmethod
+    def name_must_contain_space(cls, v: str) -> str:
+        if ' ' not in v:
+            raise ValueError('must contain a space')
+        return v.title()
+
+print(UserModel(id=1, name='John Doe'))
+#> id=1 name='John Doe'
+
+try:
+    UserModel(id=1, name='jason')
+except ValidationError as e:
+    print(e)
+    """
+    1 validation error for UserModel
+    name
+      Value error, must contain a space [type=value_error, input_value='jason', input_type=str]
+    """
+```
+
+Validation is a fundamental concept in Software 1.0. However, when we want an ai system to raise an error or self-correct, we introduce new vocabulary and terms. Instead, we can apply the same idea of having types, such as an integer or a string, but with additional constraints. For example, a string type that is not "an apology" or "a threat." The underlying principles remain the same.
+
+```python
+def validation_function(cls, value):
+    if condition(value):
+        raise ValueError("Value is not valid")
+    return mutation(value)
+```
+
+We should be able to define new types that are powered by probabilsitic models and use them in the same way we use define validators in Pydantic.
+
+## Probabilistic Validation in Software 3.0
+
+In this note, we will discuss probabilistic validation in software 3.0. We introduce a type hint called `llm_validator` and demonstrate how to create custom validators using `instructor` itself.
+
+```python
+from pydantic import BaseModel, field_validator, ValidationError
+from instruct import llm_validator
+
+class UserModel(BaseModel):
+    id: int
+    name: str
+    beliefs: str
+
+    @field_validator('beliefs')
```

Comment:
Running validation on beliefs is pretty cool. Might want to include one or two more examples of other types of validations.

Reply:
Good call on using different examples. 

I should include one that checks the chain of thought makes sense.